module: allow management via systemd socket

frantisekhanzlikbl opened the pull request:
Supersedes #52. Partially fixes #63 (per-server user & group configuration missing).

This PR adds a way to make servers managed via a systemd stdin socket and the journal, allowing for a better integrated experience with the rest of the system.

This PR largely builds on top of @Misterio77's excellent #52 and retains the nice additions:
* expanded tests verifying automated command execution
* extra lifecycle hooks for the server services

I did change the default systemd socket path from #52 to be aligned with the current tmux socket path and renamed the lifecycle hooks to preserve the systemd naming convention, though.

I've added a few extras, not all of which I'm sure want to be kept. Feedback welcome. :slightly_smiling_face:
* `docsAsciiDoc` and `docsCommonMark` package outputs (not present in the overlay) to make sure that `nix flake check` verifies that the docs can be built. there was a trivial issue previously, which prevented them from building.
* VS Code config for [nix-ide] specifying the formatter to be `nixpkgs-fmt`, along with the proper indentation.
* per-server management of the socket (both tmux & systemd) paths.

I'm not too sure I like the way server overrides of the global management system config works, but it was all I could conjure with my limited knowledge of advanced usage of the module system.

This PR should be backwards compatible, although some field-testing would be much appreciated as I don't have any real setup to test it on (and frankly no time either :)).

Infinidoge commented on modules/minecraft-servers.nix:
```diff
@@ -101,13 +212,6 @@ in
       the server name in this directory, such as <literal>/srv/minecraft/servername</literal>.
     '';
 
-    runDir = mkOpt' types.path "/run/minecraft" ''
-      Directory to place the runtime tmux sockets into.
-      Each server's console will be a tmux socket file in the form of <literal>servername.sock</literal>.
-      To connect to the console, run `tmux -S /run/minecraft/servername.sock attach`,
-      press `Ctrl + b` then `d` to detach.
-    '';
-
```
For the sake of backwards compatibility, this option should be kept, with a warning if it is changed from the default. (And it should accordingly be used in the default tmux socket path function.

Infinidoge commented on modules/minecraft-servers.nix:
```diff
@@ -179,8 +297,6 @@ in
 
           restart = mkOpt' types.str "always" ''
             Value of systemd's <literal>Restart=</literal> service configuration option.
-            Due to the servers being started in tmux sockets, values other than
-            <literal>"no"</literal> and <literal>"always"</literal> may not work properly.
```
This should be edited to maintain the note that values other than `no` and `always` may not work properly with tmux sockets.

Infinidoge reviewed (changes requested):
Overall the implementation looks good! Left some notes with regards to backwards compatibility and documentation. I'll test and do further review on the code later when I have more time.

With that said, something that you should do before this can be merged is split this up into commits with separate concerns. Namely:
- The vscode settings file should be a meta commit
- Documentation export should be its own commit
- Extra pre/post start/stop and custom stop command should be its own commit
- Adding/modifying the tests should be their own commits

frantisekhanzlikbl commented:
Thanks so much for the review!

I just pushed the requested changes. Please let me know if there is something else to change.

frantisekhanzlikbl commented:
Hmm, I just tried to migrate my config to this and got bitten in my ass by the extra lifecycle hooks not having `set -euo pipefail`, therefore "succeeding" even if parts of them fail. Would you mind me using `writeShellApplication` instead of `writeShellScript` for them?

This would theoretically be a breaking change, as setups previously "working" might now start failing, but I'd argue that these were broken from the start. I could add an option for that, but I don't think it would be worth the extra code and maintenance.

Infinidoge commented on modules/minecraft-servers.nix:
```diff
@@ -102,10 +213,12 @@ in
     '';
 
     runDir = mkOpt' types.path "/run/minecraft" ''
-      Directory to place the runtime tmux sockets into.
+      Deprecated: Directory to place the runtime tmux sockets into.
       Each server's console will be a tmux socket file in the form of <literal>servername.sock</literal>.
       To connect to the console, run `tmux -S /run/minecraft/servername.sock attach`,
       press `Ctrl + b` then `d` to detach.
+
+      Plase use <option>services.minecraft-servers.managementSystem.tmux.socketPath</option>` instead.
```
This should have a corresponding [module system warning](https://nixos.org/manual/nixos/unstable/#sec-assertions). This would work:
```
warnings = lib.optional (cfg.runDir != "/run/minecraft") ''
  `runDir` has been deprecated.

  Please use `services.minecraft-servers.managementSystem.tmux.socketPath` instead.
  For example, `name: "${cfg.runDir}/''${name}.sock"`

  See the changelog file for more information.
'';
```
Otherwise end users who don't check this repo every so often won't get the message.

frantisekhanzlikbl replied:
ouch, I totally forgot about that. sorry, I'll add it shortly. :sweat_smile:

Infinidoge commented on modules/minecraft-servers.nix:
```diff
@@ -72,6 +72,117 @@ let
+          description = ''
+            Function from a server name to the path at which the server's tmux socket is placed.
+            To connect to the console, run `tmux -S <path to socket> attach`,
+            press `Ctrl + b` then `d` to detach.
```
Could you add a note here that the default will change to be `name: "/run/minecraft/${name}.sock`" after `runDir` gets removed?

Infinidoge commented on modules/minecraft-servers.nix:
```diff
@@ -72,6 +72,117 @@ let
+            description = ''
+              Function from a server name to the path at which the server's stdin socket is placed.
+              You can send the server commands by writing to this socket,
+              for example with shell redirection: `echo 'list' > <path to socket>`.
```
Same as previous comment, would like a note stating that the default will be changing, just so its documented.

Infinidoge commented on modules/minecraft-servers.nix:
```diff
@@ -72,6 +72,117 @@ let
+      ms = server.managementSystem;
+      tmux = "${getBin pkgs.tmux}/bin/tmux";
+    in
+    assert assertMsg (!(ms.tmux.enable && ms.systemd-socket.enable)) "Only one server management system can be enabled at a time.";
```
This should be done with a [module system assertion](https://nixos.org/manual/nixos/unstable/#sec-assertions) instead. Which actually, I'm pretty sure can be handled in the `managementSystem` submodule type, which would be clean.

frantisekhanzlikbl replied:
I thought about that, but for some reason decided against it. I honestly don't remember why, and if it works at the submodule level, I don't see why not.

frantisekhanzlikbl replied:
Oh, I know. Doing the validation as a module system assertion at the submodule level prevents the user from having `{ tmux.enable = false; systemd-socket.enable = false; }` in the global config and then choosing for each server. It might be too niche of a use case, but I can very much see myself doing this.

Infinidoge replied:
Makes sense. It feels unclean, but doesn't really need to be perfect, and it works as is.

Infinidoge commented on modules/minecraft-servers.nix:
```diff
@@ -72,6 +72,117 @@ let
+            ''}
+          '';
+        };
+      } else builtins.throw "At least one server management system must be enabled.";
```
Same as previous, this would be best handled with a module system assertion.

frantisekhanzlikbl replied:
Even if this one is handled via the module system assertions, this throw must stay here too, as we really don't have any meaningful return value if no management system is enabled and returning an arbitrary one feels really wrong.

Infinidoge replied:
The module system assertion would force at least 1 management system to be enabled, since it would refuse to finish evaluating otherwise.

frantisekhanzlikbl replied:
I believe that is only true for evaluating entire NixOS configurations. i.e. things that evaluate the top-level system attribute[^1]. Any uses of the module system that do not need to do that would not inform the user of anything, and just silently produce nonsense. a good example would be manually evaluating a part of a NixOS config in the REPL during debugging, which I do fairly often.

[^1]: https://github.com/NixOS/nixpkgs/blob/db4bd243ca966a89a2bead2f9275d9163268d856/nixos/modules/system/activation/top-level.nix#L66

Infinidoge replied:
Fair. We'll just leave it as is then/

Infinidoge reviewed (changes requested):
The single period change in CHANGELOG looks out of place on the diff, but also I don't think it's big enough to warrant its own commit so it can stay.

Something that I'm a bit mixed about is how which system to use is selected. Needing to explicitly turn off the default and turn on the one you want feels... wrong to me, so an enum option like
```nix
{
  system = "tmux"
  tmux.socketPath = name: ...;
}
```
might be better. This would also alleviate the need for assertions to prevent selecting 2 at the same time, with the cost of it being not idiomatic. If anyone has any thoughts about this, please chime in and tell me my code conciseness perfectionism is on the loose again.

With regards to pipefail, yeah go right ahead. These kinds of scripts should fail early, and is the kind of breaking change that should lead to more fixed problems than actual problems.

Some other small comments about module assertions and docs. Haven't tested it myself yet, but will review again after I do.

frantisekhanzlikbl commented:
Alright, I just pushed the next iteration. Sorry about the dot change in the changelog, that might've been a little too pedantic of me. I updated the descriptions of socket path options and moved to `writeShellApplication`.

I thought about using a string enum for the preferred management system too, but decided against it, as it would make it non-practical to add more management systems (such as RCON), that don't conflict with others. i.e. both systemd-socket and tmux could be used along with RCON, just not with each other.

I kept the language-level assertions for now, for the reasons described in my comment on the appropriate review comments, but I don't feel that strongly about the first one. If you do, I can change that.

Infinidoge commented on modules/minecraft-servers.nix:
```diff
@@ -323,6 +499,15 @@ in
         }
       ];
 
+      warnings = lib.optional (cfg.runDir != options.services.minecraft-servers.runDir.default) ''
```
Hmm. How does this interact with `mkDefault`? If someone used `mkDefault`, then this might not warn them of the change.

frantisekhanzlikbl replied:
I believe that `mkDefault` is just a normal value just like any other for the module system and does not affect the default at all. There doesn't seem to be anything special about its implementation: https://github.com/NixOS/nixpkgs/blob/bc3cfcb6ed2167858598d316683f1d40ca322898/lib/modules.nix#L1014-L1020

Infinidoge replied:
Ah yep, just an override with priority, nevermind!

frantisekhanzlikbl commented:
I just remembered that systemd has its own options for timeouts, so I pushed a new commit that lets systemd handle server stop timeouts instead of the ad-hoc implementation that was in place previously in the systemd-socket management system (and completely missing in the tmux one).

marek-rychly commented:
What is the status of this merge request? Can it be merged?

Logging into journald is much more better than tmux for monitoring a server, so I am using the last commit from this merge request now. Unfortunately, there are newer fabric versions in the master branch which are not available here. Rebase or merge to the master would be appreciated. Thank you.

frantisekhanzlikbl commented:
> What is the status of this merge request? Can it be merged?
> 
> Logging into journald is much more better than tmux for monitoring a server, so I am using the last commit from this merge request now. Unfortunately, there are newer fabric versions in the master branch which are not available here. Rebase or merge to the master would be appreciated. Thank you.

I believe all the brought up concerns have been resolved on my side.

I've been using the latest commit here on a few small-scale servers in production for about three months and have faced no problems, but my deployments are also probably fairly simple.

Currently, I'm severely short on time, so if @Infinidoge has more change requests before this is mergeable, the PR will most likely have to wait for a while or get taken up by somebody else. I'll do a rebase in the meantime, though.

Infinidoge commented on modules/minecraft-servers.nix:
```diff
@@ -72,6 +72,112 @@ let
+
+            ${tmux} -S ${sock} send-keys ${escapeShellArg server.stopCommand} Enter
+
+            while server_running; do sleep 1s done
```
```suggestion
            while server_running; do sleep 1s; done
```

frantisekhanzlikbl replied:
ouch, dumb mistake. sorry and thanks for catching it. :)

Infinidoge replied:
No problem :)

I should definitely add CI since this could have been caught sooner by just having GH run `nix flake check`. It's on my I'll Do It Eventually list

Infinidoge commented:
Note to self: GitHub action that runs `nix flake check`

frantisekhanzlikbl commented:
thanks for the reviews and merge, it is greatly appreciated! :)

Misterio77 commented:
Thanks so much for picking this up! Looks awesome.
